### PR TITLE
Add option to unfuse Wqkv

### PR DESCRIPTION
**Pull request by snarayan21**
Adding an option to unfuse Wqkv. Gives us flexibility to support various other model architectures that don't fuse Wqkv like we do.

Run with fused Wqkv (main): [fused-wqkv-orig-n7eky6](https://wandb.ai/mosaic-ml/bigboi-development/runs/r7o24fax)
Run with fused Wqkv (this branch): [fused-wqkv-new-ggXoVL](https://wandb.ai/mosaic-ml/bigboi-development/runs/sqwrmja6)
Run without fused Wqkv: [unfused-wqkv-new-a2Qo2r](https://wandb.ai/mosaic-ml/bigboi-development/runs/md1t30vf)

Same loss curves:
Screenshot 2024-07-17 at 11 39 08 AM

Same throughput for fused Wqkv before and after, slightly lower throughput for unfused Wqkv (as expected):
Screenshot 2024-07-17 at 11 38 42 AM

### CODE REVIEW

File: llmfoundry/models/layers/attention.py [hunk 1]
```diff
@@ -462,29 +464,72 @@ def __init__(
...
+                    out_features=self.kv_n_heads * self.head_dim,
+                    fc_kwargs=fc_type,
+                )
+                self.Wv = build_fc(
+                    name=fc_type_name,
+                    in_features=self.d_model,
+                    out_features=self.kv_n_heads * self.head_dim,
+                    fc_kwargs=fc_type,
+                )
+                # for param init fn; enables shape based init of fused layers
```
Comment (irenedea) [inline llmfoundry/models/layers/attention.py:514]: do we need this if this is for unfused?

File: llmfoundry/models/layers/attention.py [hunk 2]
```diff
@@ -411,6 +411,7 @@ def __init__(
         clip_qkv: Optional[float] = None,
         qk_ln: bool = False,
         qk_gn: bool = False,
+        fused_weights: bool = True,
```
Comment (dakinggg) [inline llmfoundry/models/layers/attention.py:414]: seems me to this should just be `fused_qkv`, right?

File: llmfoundry/models/layers/attention.py [hunk 3]
```diff
@@ -462,29 +464,72 @@ def __init__(
...
-                out_features=self.d_model + 2 * self.kv_n_heads * self.head_dim,
-                fc_kwargs=fc_type,
-            )
-            # for param init fn; enables shape based init of fused layers
-            fuse_splits = [
-                i * self.head_dim
-                for i in range(1, self.n_heads + 2 * self.kv_n_heads)
-            ]
-            self.Wqkv._fused = (0, fuse_splits)
+        if self.fused_weights:
```
Comment (dakinggg) [inline llmfoundry/models/layers/attention.py:467]: I think this code can be simplified a bit, currently a bunch of duplication. basically `if reuse_kv_layer_idx`, we're always making Wq, and then if fused we make wqkv, else we make wk and wv

**Review by dakinggg — COMMENTED** [review]
please add some unit tests too

File: llmfoundry/models/layers/attention.py (hunk 3 above)
Comment (snarayan21) [reply llmfoundry/models/layers/attention.py:467]: yea wanted to get a branch up for quick testing. sg.

File: llmfoundry/models/layers/attention.py (hunk 2 above)
Comment (snarayan21) [reply llmfoundry/models/layers/attention.py:414]: sure

File: llmfoundry/models/layers/attention.py (hunk 3 above)
Comment (snarayan21) [reply llmfoundry/models/layers/attention.py]: done

File: llmfoundry/models/layers/attention.py [hunk 4]
```diff
@@ -600,20 +626,29 @@ def get_qkv(
                 dtype = query.dtype
                 query = self.q_ln(query).to(dtype).view(q_shape)
             return query, key, value
+        elif self.fused_qkv:
```
Comment (dakinggg) [inline llmfoundry/models/layers/attention.py:629]: this should just be separate if statement i think since the above if statement returns

**Review by dakinggg — COMMENTED** [review]
lgtm, would like @vchiley or @ShashankMosaicML to take a look before approving

File: llmfoundry/models/layers/attention.py (hunk 4 above)
Comment (snarayan21) [reply llmfoundry/models/layers/attention.py]: done

**Comment by snarayan21** [comment]
Added testing for fuse splits into the unit test @dakinggg

**Review by ShashankMosaicML — APPROVED** [review]
lgtm.

**Review by dakinggg — APPROVED** [review]

**Pull request merged**